File: verdict-backend/src/routes/sessions.ts
import type { FastifyInstance, FastifyRequest, FastifyReply } from 'fastify';
import { z } from 'zod';
import { db } from '../lib/db.js';
import { requireAuth } from '../middleware/auth.js';
import { generateQuestion } from '../agents/interrogator.js';
import { analyzeForObjections } from '../agents/objection.js';
import { detectInconsistency } from '../agents/detector.js';

const createSessionBody = z.object({
  durationMinutes: z.union([z.literal(15), z.literal(30), z.literal(45), z.literal(60)]),
  focusAreas: z.array(z.string()).min(1).max(6),
  aggression: z.enum(['STANDARD', 'ELEVATED', 'HIGH_STAKES']),
  objectionCopilotEnabled: z.boolean(),
  sentinelEnabled: z.boolean(),
});

export async function sessionsRoutes(app: FastifyInstance) {
  app.get<{ Params: { sessionId: string } }>(
    '/:sessionId',
    { preHandler: [requireAuth] },
    async (request, reply) => {
      const user = request.user!;
      const { sessionId } = request.params;
      const session = await db.session.findFirst({
        where: { id: sessionId, firmId: user.firmId },
      });
      if (!session)
        return reply.code(404).send({ success: false, error: { code: 'NOT_FOUND', statusCode: 404 } });
      return reply.send({
        success: true,
        data: {
          id: session.id,
          status: session.status,
          sessionNumber: session.sessionNumber,
          durationMinutes: session.durationMinutes,
          startedAt: session.startedAt,
          questionCount: session.questionCount,
        },
      });
    }
  );

  app.post<{ Params: { sessionId: string } }>(
    '/:sessionId/start',
    { preHandler: [requireAuth] },
    async (request, reply) => {
      const user = request.user!;
      const { sessionId } = request.params;
      const session = await db.session.findFirst({
        where: { id: sessionId, firmId: user.firmId },
      });
      if (!session)
        return reply.code(404).send({ success: false, error: { code: 'NOT_FOUND', statusCode: 404 } });
      await db.session.update({
        where: { id: sessionId },
        data: { status: 'ACTIVE', startedAt: new Date() },
      });
      return reply.send({
        success: true,
        data: { sessionId, status: 'ACTIVE', startedAt: new Date().toISOString() },
      });
    }
  );

  app.post<{ Params: { sessionId: string } }>(
    '/:sessionId/agents/question',
    { preHandler: [requireAuth] },
    async (request, reply) => {
      const user = request.user!;
      const { sessionId } = request.params;
      const body = request.body as {
        questionNumber: number;
        priorAnswer?: string;
        hesitationDetected?: boolean;
        recentInconsistencyFlag?: boolean;
        currentTopic?: string;
      };
      const session = await db.session.findFirst({
        where: { id: sessionId, firmId: user.firmId },
        include: { case: true },
      });
      if (!session)
        return reply.code(404).send({ success: false, error: { code: 'NOT_FOUND', statusCode: 404 } });

      reply.raw.setHeader('Content-Type', 'text/event-stream');
      reply.raw.setHeader('Cache-Control', 'no-cache');
      reply.raw.setHeader('Connection', 'keep-alive');
      let fullText = '';
      try {
        reply.raw.write(
          `data: ${JSON.stringify({ type: 'QUESTION_START', questionNumber: body.questionNumber ?? 1 })}\n\n`
        );
        const stream = generateQuestion({
          caseType: session.case.caseType,
          witnessRole: 'DEFENDANT',
          currentTopic: body.currentTopic ?? 'PRIOR_STATEMENTS',
          aggressionLevel: (session.aggression as 'STANDARD' | 'ELEVATED' | 'HIGH_STAKES') ?? 'STANDARD',
          priorAnswer: body.priorAnswer,
          questionNumber: body.questionNumber ?? 1,
          hesitationDetected: body.hesitationDetected ?? false,
          recentInconsistencyFlag: body.recentInconsistencyFlag ?? false,
          niaSessionContextId: session.niaSessionContextId ?? session.id,
          priorWeakAreas: [],
        });
        for await (const chunk of stream) {
          fullText += chunk;
          reply.raw.write(`data: ${JSON.stringify({ type: 'QUESTION_CHUNK', text: chunk })}\n\n`);
        }
        reply.raw.write(
          `data: ${JSON.stringify({ type: 'QUESTION_END', fullText, latencyMs: 0 })}\n\n`
        );
      } finally {
        reply.raw.end();
      }
    }
  );

  app.post<{ Params: { sessionId: string } }>(
    '/:sessionId/agents/objection',
    { preHandler: [requireAuth] },
    async (request, reply) => {
      const user = request.user!;
      const { sessionId } = request.params;
      const body = request.body as { questionNumber: number; questionText: string };
      const session = await db.session.findFirst({
        where: { id: sessionId, firmId: user.firmId },
      });
      if (!session)
        return reply.code(404).send({ success: false, error: { code: 'NOT_FOUND', statusCode: 404 } });
      const start = Date.now();
      const result = await analyzeForObjections({
        questionText: body.questionText,
        sessionId,
      });
      const processingMs = Date.now() - start;
      return reply.send({
        success: true,
        data: { ...result, processingMs },
      });
    }
  );

  app.post<{ Params: { sessionId: string } }>(
    '/:sessionId/agents/inconsistency',
    { preHandler: [requireAuth] },
    async (request, reply) => {
      const user = request.user!;
      const { sessionId } = request.params;
      const body = request.body as {
        questionNumber: number;
        questionText: string;
        answerText: string;
      };
      const session = await db.session.findFirst({
        where: { id: sessionId, firmId: user.firmId },
      });
      if (!session)
        return reply.code(404).send({ success: false, error: { code: 'NOT_FOUND', statusCode: 404 } });
      const result = await detectInconsistency({
        questionText: body.questionText,
        answerText: body.answerText,
        sessionId,
        niaSessionContextId: session.niaSessionContextId ?? session.id,
        caseType: 'MEDICAL_MALPRACTICE',
      });
      return reply.send({ success: true, data: result });
    }
  );
}

File: verdict-backend/app/routers/sessions.py
import json
import time
import base64
from datetime import datetime
from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import StreamingResponse
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from sqlalchemy.orm import selectinload
from app.database import get_db
from app.middleware.auth import require_auth
from app.models.user import User
from app.models.session import Session
from app.agents.interrogator import generate_question
from app.agents.objection import analyze_for_objections
from app.agents.detector import detect_inconsistency
from app.services.elevenlabs import text_to_speech
from app.schemas.sessions import QuestionRequest, ObjectionRequest, InconsistencyRequest
from app.config import settings

router = APIRouter()


@router.get("/{session_id}")
async def get_session(session_id: str, db: AsyncSession = Depends(get_db), user: User = Depends(require_auth)):
    result = await db.execute(
        select(Session).where(Session.id == session_id, Session.firm_id == user.firm_id)
    )
    session = result.scalar_one_or_none()
    if not session:
        raise HTTPException(404, detail={"code": "NOT_FOUND"})
    return {
        "success": True,
        "data": {
            "id": session.id,
            "status": session.status,
            "sessionNumber": session.session_number,
            "durationMinutes": session.duration_minutes,
            "startedAt": session.started_at,
            "questionCount": session.question_count,
        },
    }


@router.post("/{session_id}/start")
async def start_session(session_id: str, db: AsyncSession = Depends(get_db), user: User = Depends(require_auth)):
    result = await db.execute(
        select(Session).where(Session.id == session_id, Session.firm_id == user.firm_id)
    )
    session = result.scalar_one_or_none()
    if not session:
        raise HTTPException(404, detail={"code": "NOT_FOUND"})
    session.status = "ACTIVE"
    session.started_at = datetime.utcnow()
    await db.commit()
    return {"success": True, "data": {"sessionId": session_id, "status": "ACTIVE", "startedAt": session.started_at.isoformat()}}


@router.post("/{session_id}/agents/question")
async def stream_question(
    session_id: str,
    body: QuestionRequest,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(require_auth),
):
    result = await db.execute(
        select(Session)
        .where(Session.id == session_id, Session.firm_id == user.firm_id)
        .options(selectinload(Session.case))
    )
    session = result.scalar_one_or_none()
    if not session:
        raise HTTPException(404, detail={"code": "NOT_FOUND"})

    async def event_stream():
        full_text = ""
        yield f"data: {json.dumps({'type': 'QUESTION_START', 'questionNumber': body.questionNumber})}\n\n"

        async for chunk in generate_question(
            case_type=session.case.case_type if session.case else "OTHER",
            witness_role="DEFENDANT",
            current_topic=body.currentTopic,
            aggression_level=session.aggression or "STANDARD",
            nia_session_context_id=session.nia_session_context_id or session.id,
            question_number=body.questionNumber,
            prior_answer=body.priorAnswer,
            hesitation_detected=body.hesitationDetected,
            recent_inconsistency_flag=body.recentInconsistencyFlag,
            prior_weak_areas=[],
        ):
            full_text += chunk
            yield f"data: {json.dumps({'type': 'QUESTION_CHUNK', 'text': chunk})}\n\n"

        try:
            audio = await text_to_speech(full_text, settings.ELEVENLABS_INTERROGATOR_VOICE_ID)
            yield f"data: {json.dumps({'type': 'QUESTION_AUDIO', 'audioBase64': base64.b64encode(audio).decode()})}\n\n"
        except Exception:
            pass

        yield f"data: {json.dumps({'type': 'QUESTION_END', 'fullText': full_text})}\n\n"

    return StreamingResponse(event_stream(), media_type="text/event-stream")


@router.post("/{session_id}/agents/objection")
async def check_objection(
    session_id: str,
    body: ObjectionRequest,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(require_auth),
):
    result = await db.execute(
        select(Session).where(Session.id == session_id, Session.firm_id == user.firm_id)
    )
    if not result.scalar_one_or_none():
        raise HTTPException(404, detail={"code": "NOT_FOUND"})

    start = time.time()
    analysis = await analyze_for_objections(question_text=body.questionText, session_id=session_id)
    return {"success": True, "data": {**analysis, "processingMs": int((time.time() - start) * 1000)}}


@router.post("/{session_id}/agents/inconsistency")
async def check_inconsistency(
    session_id: str,
    body: InconsistencyRequest,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(require_auth),
):
    result = await db.execute(
        select(Session)
        .where(Session.id == session_id, Session.firm_id == user.firm_id)
        .options(selectinload(Session.case))
    )
    session = result.scalar_one_or_none()
    if not session:
        raise HTTPException(404, detail={"code": "NOT_FOUND"})

    detection = await detect_inconsistency(
        question_text=body.questionText,
        answer_text=body.answerText,
        session_id=session_id,
        nia_session_context_id=session.nia_session_context_id or session_id,
        case_type=session.case.case_type if session.case else "OTHER",
    )
    return {"success": True, "data": detection}
